Userland/Libraries/LibGfx/TextLayout.h

```diff
@@ -75,49 +75,21 @@ enum class IncludeLeftBearing {
     No
 };
 
-struct GlyphPosition {
-    FloatPoint position;
-    float glyph_width;
-    AK::Utf8CodePointIterator& it;
+struct DrawGlyph {
+    FloatPoint point;
+    u32 code_point;
+    Font const* font;
 };
 
-template<typename Callback>
-void for_each_glyph_position(FloatPoint start, Utf8View text, Font const& font, Callback callback, IncludeLeftBearing include_left_bearing = IncludeLeftBearing::No)
-{
-    float space_width = font.glyph_width(' ') + font.glyph_spacing();
-
-    u32 last_code_point = 0;
-
-    auto point = start;
-    point.translate_by(0, -font.pixel_metrics().ascent);
-
-    for (auto code_point_iterator = text.begin(); code_point_iterator != text.end(); ++code_point_iterator) {
-        auto code_point = *code_point_iterator;
-        if (should_paint_as_space(code_point)) {
-            point.translate_by(space_width, 0);
-            last_code_point = code_point;
-            continue;
-        }
-
-        auto kerning = font.glyphs_horizontal_kerning(last_code_point, code_point);
-        if (kerning != 0.0f)
-            point.translate_by(kerning, 0);
-
-        auto it = code_point_iterator; // The callback function will advance the iterator, so create a copy for this lookup.
-        auto glyph_width = font.glyph_or_emoji_width(it) + font.glyph_spacing();
-
-        auto glyph_point = point;
-        if (include_left_bearing == IncludeLeftBearing::Yes)
-            glyph_point += FloatPoint(font.glyph_left_bearing(code_point), 0);
+struct DrawEmoji {
+    IntPoint point;
+    Gfx::Bitmap const* emoji;
+    Font const* font;
+};
 
-        callback(GlyphPosition {
-            .position = glyph_point,
-            .glyph_width = glyph_width,
-            .it = code_point_iterator });
+using DrawGlyphOrEmoji = Variant<DrawGlyph, DrawEmoji>;
 
-        point.translate_by(glyph_width, 0);
-        last_code_point = code_point;
-    }
-}
+Variant<DrawGlyph, DrawEmoji> prepare_draw_glyph_or_emoji(FloatPoint point, Utf8CodePointIterator& it, Font const& font);
+Vector<DrawGlyphOrEmoji> get_glyph_run(FloatPoint baseline_start, Utf8View const& string, Font const& font, IncludeLeftBearing include_left_bearing = IncludeLeftBearing::No);
```

Comment:
Could this be passed a callback rather than return a vector? None if the callers in LibGfx want a vector (as they iterate over it immediately).

```c++
void for_each_glyph_in_run(...., Function<void(DrawGlyphOrEmoji)>)

Vector<DrawGlyphOrEmoji> get_glyph_run(...) {
    Vector<DrawGlyphOrEmoji> glyph_or_emoijs;
    for_each_glyph_in_run(..., [&](auto glyph_or_emoji) { glyph_or_emoijs.append(glyph_or_emoji); });
}
```

Reply:
sure, I added `for_each_glyph_in_run` for callers in LibGfx and made `get_glyph_run` use it.